pkg/resource/deploy/source_eval.go

```diff
@@ -150,9 +151,23 @@ func (src *evalSource) Iterate(ctx context.Context, providers ProviderSource) (S
 	regChan := make(chan *registerResourceEvent)
 	regOutChan := make(chan *registerResourceOutputsEvent)
 	regReadChan := make(chan *readResourceEvent)
+	finChan := make(chan error)
+	// We receive from programCompleteChan during shutdown. This channel needs
+	// to be buffered because a program can exit before we are shutting down.
+	programCompleteChan := make(chan error, 1)
```

Comment:
If this can only be sent to once the code might be easier using our promise library?

Reply:
Yes, thanks for the suggestion, that is tidier!